### PR TITLE
On-demand SetupAllProperties

### PR DESCRIPTION
**Pull request by vanashimko**
It is implementation of lazy `SetupAllProperties`, as described in #740 by @stakx.

New switch `AutoSetupProperties` is added, current `SetupAllProperties` implementation is slightly hacked to behave the same as it was before. Eager setup is moved to interceptor to setup properties being accessed on-demand.

New code passed all existing tests and I've added some more based on comments in the existing code about corner cases.

As the word "switch" was mentioned in the discussion, I decided to go via existing `Switches` enum. What bothers me is that now there are two ways of doing *almost* the same: `SetupAllProperties` call as before and enabling `AutoSetupProperties` switch directly. `SetupAllProperties` is hacked to behave like eager version before (by saving `DefaultValueProvider`) and switch doesn't have this (I just decided so, but we can actually save `DefaultValueProvider` in the setter of `Switches`...). It looks logically to me, that switch does not save `DefaultValueProvider`, since switch is more declarative, while `SetupAllProperties` is imperative (even though it is implemented in a new way).

What do you think?

### CODE REVIEW

**Comment by stakx** [comment]
Hi @vanashimko, I likely won't have time to review your PR until sometime in June due to being very busy this month, but I'll eventually get to it. I'd like to ask you to be patient, though.

It would've been good to discuss this in an issue first, since there are some finer (and AFAIK undocumented) points you might not yet be aware of which, if we're unlucky, make the whole idea of a lazy `SetupAllProperties` non-feasible (we might currently be lacking unit tests for that). I'm mainly thinking of correct choice of method overloads + setup order here... it's a pain point in Moq that hasn't been solved cleanly yet, and it was a very central issue the last time I had to touch `SetupAllProperties`.

That being said, judging from your description, you definitely seem to be on the correct track here. Ideally, `SetupAllProperties` should do nothing more than enable the switch. Using `Switches` to house that new switch is sensible (although I'd hoped to deprecate `Switches` in the near future). Adding a new step in the interception pipeline (before `Return`) is what I'd have done, too.

I'll try to review in detail soon.

**Comment by vanashimko** [comment]
Thanks for you response, @stakx!

Even if you'll find that due to some "finer points" lazy `SetupAllProperties` is non-feasible, it was a nice time hacking around the codebase...

**Comment by stakx** [comment]
To give some more context for the ominous "finer points" (should you choose to look into it):

You might have noticed code that gathers all properties in topological ("depth-first") order. I only took a quick look at your first commit and saw that you removed this logic (and I'm glad you did, as it's part of why `SetupAllProperties` has bad performance). In the absence of that code, it's now important to ensure that more derived properties are still favoured over less derived ones when choosing which one to set up (e.g. for the case where a derived class makes a read-only base property writable, too; and there may be more cases such as properties from a reimplemented interface). When property stubbing becomes lazy, I'm not sure this is still guaranteed as the order in which you access properties might now start influencing things and we have no control over that.

I suspect that things still work (i.e. all tests pass) despite the removal of the topological search due to the recently introduced `GetReboundProperty` helper method, but that's just a suspicion; we should take a closer look at this to ensure we don't break anything.

**Comment by vanashimko** [comment]
Thanks, I'll try to get a closer look and add more tests. 

My thoughts were that if we already have calling method (which represents accessing property) in `Invocation` passed to `IInterceptor`, we should not bother much about the order of traversal of hierarchy. The were some stuff related to base interfaces (`GetProperty` surprisingly does not return property of base interfaces if called on an interface) and reimplemented interfaces via `As`, but there are tests for this case, so I catched this at early stage.

If there are some nifty moments with the method we have in `Invocation.Method`, so maybe my optimistic assumption was wrong...

File: src/Moq/Interception/InterceptionAspects.cs [hunk 1]
```diff
@@ -331,4 +332,131 @@ public static void Handle(Invocation invocation, Mock mock)
...
+		{
+			MethodInfo invocationMethod = invocation.Method;
+			// Original implementation with eager setup skipped indexers too.
+			if (invocationMethod.IsPropertyAccessor() && !invocationMethod.IsPropertyIndexerAccessor())
+			{
+				if (mock.Setups.FindMatchFor(invocation) == null)
+				{
+					PropertyInfo property = GetPropertyFromAccessorMethod(mock, invocationMethod, out Type propertyHolderType);
+
+					// Should ignore write-only properties, as they will be handled by Return aspect.
```
Comment (stakx) [inline src/Moq/Interception/InterceptionAspects.cs]: ... unless if the mock is strict; for those, interception terminates even before we get to `Return`!

**Comment by vanashimko** [comment]
@stakx, thanks for the so detailed review! I'll be looking into the comments.

File: src/Moq/Interception/InterceptionAspects.cs [hunk 2]
```diff
@@ -331,4 +332,117 @@ public static void Handle(Invocation invocation, Mock mock)
...
+							try
+							{
+								initialValue = mock.GetDefaultValue(getter, out innerMock,
+									useAlternateProvider: mock.DefaultValueProvider);
+							}
+							catch
+							{
+								// Since this method performs a batch operation, a single failure of the default value
+								// provider should not tear down the whole operation. The empty default value provider
+								// is a safe fallback because it does not throw.
```
[the remark this answers is not in the document]
Comment (vanashimko) [reply src/Moq/Interception/InterceptionAspects.cs]: Sorry, didn't get it... 
`SetupProperty` just uses `default(T)`, which I suspect equivalent to `DefaultValueProvider.Empty`:
https://github.com/moq/moq4/blob/9ea157cf61f1eb7deeadc419cac08d1aa9fd4f44/src/Moq/Mock.Generic.cs#L302-L325

But here we're using value provider, which can throw. Did you mean that we should allow value provider to throw and do not catch any exceptions?

File: src/Moq/Interception/InterceptionAspects.cs [hunk 3]
```diff
@@ -380,7 +380,7 @@ public static void Handle(Invocation invocation, Mock mock)
 
 							if (innerMock != null)
 							{
-								Mock.SetupAllProperties(innerMock);
+								Mock.SetupAllProperties(innerMock, mock.AutoSetupPropertiesDefaultValueProvider);
```
[the remark this answers is not in the document]
Comment (vanashimko) [reply src/Moq/Interception/InterceptionAspects.cs]: I did it this way because I thought that the following lines
https://github.com/moq/moq4/blob/9ea157cf61f1eb7deeadc419cac08d1aa9fd4f44/src/Moq/Mock.cs#L553-L556

does the same in current implementation.

`defaultValueProvider` passed from
https://github.com/moq/moq4/blob/9ea157cf61f1eb7deeadc419cac08d1aa9fd4f44/src/Moq/Mock.cs#L497
is captured for inner mocks here.

That was my understanding. What did I miss here?

File: src/Moq/Mock.cs [hunk 4]
```diff
@@ -495,6 +501,13 @@ private static TSetup SetupRecursive<TSetup>(Mock mock, LambdaExpression express
 		internal static void SetupAllProperties(Mock mock)
 		{
 			mock.Switches |= Switches.AutoSetupProperties;
+
+			mock.AutoSetupPropertiesDefaultValueProvider = mock.DefaultValueProvider;
+			// `SetupAllProperties` no longer performs property setup like in previous versions.
+			// Instead it just enables a switch to setup properties on-demand at the moment of first access.
+			// In order for `SetupAllProperties`'s new mode of operation to be indistinguishable
+			// from how it worked previously, it's important to capture the default value provider at this precise
+			// moment, since it might be changed later (before queries to properties).
```
[the remark this answers is not in the document]
Comment (vanashimko) [reply src/Moq/Mock.cs]: Agree, really like your solution! Will remove new switch with pleasure 😄

File: tests/Moq.Tests/StubExtensionsFixture.cs [hunk 5]
```diff
@@ -149,6 +149,20 @@ public void Property_stubbed_by_SetupAllProperties_during_DefaultValue_Mock_can_
...
+			public abstract string Test { set; }
+		}
+
+		[Fact]
+		public void Write_only_property_should_be_ignored_by_SetupAllProperties()
+		{
+			var mock = new Mock<WriteOnlyProperty>();
+			mock.SetupAllProperties();
+			
+			mock.Object.Test = "test";
```
[the remark this answers is not in the document]
Comment (vanashimko) [reply tests/Moq.Tests/StubExtensionsFixture.cs]: Good catch! If current behavior is to throw, then I would rather not change it in scope of this PR.

File: src/Moq/Interception/InterceptionAspects.cs [hunk 6]
```diff
@@ -331,4 +332,131 @@ public static void Handle(Invocation invocation, Mock mock)
...
+
+		public static void Handle(Invocation invocation, Mock mock)
+		{
+			MethodInfo invocationMethod = invocation.Method;
+			// Original implementation with eager setup skipped indexers too.
+			if (invocationMethod.IsPropertyAccessor() && !invocationMethod.IsPropertyIndexerAccessor())
+			{
+				if (mock.Setups.FindMatchFor(invocation) == null)
+				{
+					PropertyInfo property = GetPropertyFromAccessorMethod(mock, invocationMethod, out Type propertyHolderType);
```
[the remark this answers is not in the document]
Comment (vanashimko) [reply src/Moq/Interception/InterceptionAspects.cs]: Great, thanks, removed a lot of not needed code!

File: tests/Moq.Tests/StubExtensionsFixture.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (vanashimko) [reply tests/Moq.Tests/StubExtensionsFixture.cs]: I've opened #835. If you're OK with fixing it separately, please resolve this discussion.

File: src/Moq/Interception/InterceptionAspects.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (stakx) [reply src/Moq/Interception/InterceptionAspects.cs]: Apologies, I should have been more specific. I mentioned `SetupProperty` because it doesn't even try to guard against failure, but you're right, that's a bad comparison because unlike a default value provider, `default(T)` as used by `SetupProperty` cannot possibly fail in the first place.

Let me explain in a different fashion. The `catch` guard in `SetupAllProperties` is there to prevent one very specific situation: If your `mock` uses `DefaultValue.Mock` and you're calling `mock.SetupAllProperties()`, and `mock` has a property `P` of type `T` which is declared serializable but implements the serializable protocol incorrectly, then initializing `P` with a new `Mock<T>` instance can fail in DynamicProxy (which has logic to deal with serializable types). To avoid complete failure of `SetupAllProperties` because of that single property, there used to be [code](https://github.com/moq/moq4/blob/0094c40e13194f5ada1acd347f200ebeaf597050/Source/Mock.cs#L742-L757) that would essentially fall back to the `DefaultValue.Empty` (i.e. no attempt to mock `T`) in case of such failure. This logic has subsequently been refactored (by yours truly) into the `catch` clause you're seeing today.

Since `SetupAllProperties` is no longer a batch operation, faulty serializables can no longer cause the operation as a whole to fail; you'd have to access the `P` property, and I think it would be perfectly legitimate to let the DynamicProxy error about the faulty serializable type "bubble up" all the way to user code. (That would admittedly be a minute functional change that could annoy users every now and then.)

Any thoughts?

[the remark this answers is not in the document]
Comment (vanashimko) [reply src/Moq/Interception/InterceptionAspects.cs]: Interesting, now I got it. Agree that catch can be removed. As for 
> Perhaps we need a new test for this aspect.

You meant that we should test that an exception from `DefaultValueProvider.GetDefaultValue` is bubbled up here?

[the remark this answers is not in the document]
Comment (stakx) [reply src/Moq/Interception/InterceptionAspects.cs]: Ah, I was looking for that test, but couldn't remember what it was called nor where it was defined. :laughing:

> Will just change this test, since now it SHOULD throw

Perfect. The new test will essentially become a special case of the existing ` Accessing_property_of_bad_serializable_type_throws` test. Seems by throwing, Moq will become more consistent.

File: src/Moq/Interception/InterceptionAspects.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (stakx) [reply src/Moq/Interception/InterceptionAspects.cs]: >What did I miss here?

Probably nothing. It seems it is I who missed something. 😄

It is possible that the present-day implementation is faulty in that regard. I still think the root mock's provider shouldn't override inner mocks' ones. But that is probably a separate issue.

What counts more for the moment is that we're not introducing a functional change here. If your current implementation works the same way as present-day Moq (perhaps ensure that with a new test), then we can resolve this discussion.

File: src/Moq/Mock.cs [hunk 7]
```diff
@@ -494,88 +500,19 @@ private static TSetup SetupRecursive<TSetup>(Mock mock, LambdaExpression express
...
-			// `SetupAllProperties` no longer performs eager recursive property setup like in previous versions.
-			// If `mock` uses `DefaultValue.Mock`, mocked sub-objects are now only constructed when queried for
-			// the first time. In order for `SetupAllProperties`'s new mode of operation to be indistinguishable
-			// from how it worked previously, it's important to capture the default value provider at this precise
-			// moment, since it might be changed later (before queries to properties of a mockable type).
+			SetupAllProperties(mock, mock.DefaultValueProvider);
 		}
 
-		private static void SetupAllProperties(Mock mock, DefaultValueProvider defaultValueProvider)
+		internal static void SetupAllProperties(Mock mock, DefaultValueProvider defaultValueProvider)
```
[the remark this answers is not in the document]
Comment (stakx) [reply src/Moq/Mock.cs]: As discussed at the other review comment, please ignore for now. My bad.

File: src/Moq/Interception/InterceptionAspects.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (stakx) [reply src/Moq/Interception/InterceptionAspects.cs]: I just ran a quick test, your implementation acts the same way as present-day Moq. My bad, let's leave your implementation as it is now.

[the remark this answers is not in the document]
Comment (vanashimko) [reply src/Moq/Interception/InterceptionAspects.cs]: I've added `Property_stubbed_by_SetupAllProperties_should_use_parent_DefaultValue_behaviour_for_inner_mocks` to capture current behavior in tests.

File: CHANGELOG.md [hunk 8]
```diff
@@ -11,6 +11,10 @@ The format is loosely based on [Keep a Changelog](http://keepachangelog.com/en/1
 
 * Regression: Unhelpful exception message when setting up an indexer with `SetupProperty` (@stakx, #823)
 
+#### Changed
+
+* Improved performance for `Mock.Of<T>` and `mock.SetupAllProperties()` as the latter now performs property setups just-in-time, instead of as an ahead-of-time batch operation. (@vanashimko, #826)
```
Comment (stakx) [inline CHANGELOG.md]: Please add under a new `## Unreleased` heading at the top of the changelog;  we're not changing an already published version, after all. :)

**Comment by vanashimko** [comment]
Thanks @stakx! 

I've added changelog entry as you suggested. I'm OK with squashing all my commits into one automatically by GitHub during merge.

File: CHANGELOG.md (hunk 8 above)
Comment (vanashimko) [reply CHANGELOG.md]: I've placed this under existing

```
## Unreleased

#### Fixed

* Regression: Unhelpful exception message when setting up an indexer with `SetupProperty` (@stakx, #823)

#### Changed

* Improved performance for `Mock.Of<T>` and `mock.SetupAllProperties()` as the latter now performs property setups just-in-time, instead of as an ahead-of-time batch operation. (@vanashimko, #826)
```

Should I add new (the second) `## Unreleased` for my one?

Comment (stakx) [reply CHANGELOG.md]: No. You're apparently seeing an outdated version of `CHANGELOG.md`. The current version looks like this: https://github.com/moq/moq4/blob/9ea157cf61f1eb7deeadc419cac08d1aa9fd4f44/CHANGELOG.md.

Comment (vanashimko) [reply CHANGELOG.md]: My bad, rebased and updated.

**Comment by stakx** [comment]
Thanks again so much for this contribution, it's been a pleasure to review.

(Please feel free to contribute more, if you are interested and can spare the time. #740 and #835 might be good candidates now.)
